staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -0,0 +1,261 @@
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
+			existing.Versions = append(existing.Versions, value)
+		}
+
+	} else {
+		group := &apidiscoveryv2beta1.APIGroupDiscovery{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []apidiscoveryv2beta1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroups[groupName] = group
+		rdm.apiGroupNames = append(rdm.apiGroupNames, group.Name)
+
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cache.Store(&cachedGroupList{})
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	modified := false
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			modified = true
+			break
+		}
+	}
+	// If no modification was done, cache does not need to be cleared
+	if !modified {
+		return
+	}
+
+	if len(group.Versions) == 0 {
+		for i, g := range rdm.apiGroupNames {
+			if g == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			}
+		}
+		delete(rdm.apiGroups, group.Name)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cache.Store(&cachedGroupList{})
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+
+	// Reset response document so it is recreated lazily
+	rdm.cache.Store(&cachedGroupList{})
+}
+
+// Prepares the api group list for serving by converting them from map into
+// list and sorting them according to insertion order
+func (rdm *resourceDiscoveryManager) calculateAPIGroups() []apidiscoveryv2beta1.APIGroupDiscovery {
+	rdm.lock.RLock()
+	defer rdm.lock.RUnlock()
+
+	// Re-order the apiGroups by their insertion order
+	groups := []apidiscoveryv2beta1.APIGroupDiscovery{}
+	for _, v := range rdm.apiGroupNames {
+		groups = append(groups, *rdm.apiGroups[v].DeepCopy())
+	}
+	return groups
+}
+
+func (rdm *resourceDiscoveryManager) generateCache() *cachedGroupList {
+	for {
+		cacheLoad := rdm.cache.Load()
+		response := apidiscoveryv2beta1.APIGroupDiscoveryList{
+			Items: rdm.calculateAPIGroups(),
+		}
+		etag, err := calculateETag(response)
+		if err != nil {
+			klog.Errorf("failed to calculate etag for discovery document: %s", etag)
+		}
+
+		cached := &cachedGroupList{
+			cachedResponse:     response,
+			cachedResponseETag: etag,
+		}
+		if success := rdm.cache.CompareAndSwap(cacheLoad, cached); success {
+			return cached
+		}
+	}
+	return nil
+}
+
+type cachedGroupList struct {
```

Comment:
It looks like `cachedResponseETag == ""` is used to decide whether this structure has been initialized or not. Maybe add a comment to describe that? Or alternatively, add a new methods:
```
func (c cachedGrouplist) isInitialized() bool {
    return cachedResponseETag == ""
}
```

Reply:
Used the alternative approach that you described